### PR TITLE
Gradpassing

### PR DESCRIPTION
**Pull request by unconst**
-- adds a prioritized gradient queue to the axon which saves gradients.
-- adds synapse.grad which returns input gradients on request. (these are not used by the calling peer as of now.
-- adds gradient accumulation functionality to the mnist and gpt neuron
-- adds a new FFNN neuron which waits for grads before applying them
-- cleans some examples

### CODE REVIEW

File: bittensor/axon.py [hunk 1]
```diff
@@ -88,7 +106,7 @@ def check_config(config: Munch) -> Munch:
                     config to check.
         """
         config = Nucleus.check_config(config)
-        logger.info('obtaining remote ip ...')
+        logger.info('optaining remote ip ...')
```
Comment (shibshib) [inline bittensor/axon.py:109]: typo. Why did we change it to optaining? lol

File: bittensor/dendrite.py [hunk 2]
```diff
@@ -250,50 +299,110 @@ def __init__(self, neuron: bittensor_pb2.Neuron, config, keypair):
...
+        self.stats.codes[code.item()] += 1
+        if code.item() == bittensor_pb2.ReturnCode.Success:
+            # We only need to record elapsed times / bytes on success.
+            self.stats.total_success_time += elapsed_time
+            self.stats.avg_success_time = self.stats.total_success_time / self.stats.codes[bittensor_pb2.ReturnCode.Success]
+            self.stats.success_bytes_out += sys.getsizeof(inputs)
+            self.stats.success_bytes_in += sys.getsizeof(outputs)
+
+        # ---- On Success: set zero backoff and halve the next backoff ---- 
+        if code.item() == bittensor_pb2.ReturnCode.Success:
```
Comment (shibshib) [inline bittensor/dendrite.py:354]: Guessing you rebased this off the #114  change? :) same comment applies, not sure why we're checking twice for the same return code.

File: bittensor/dendrite.py [hunk 3]
```diff
@@ -343,7 +452,7 @@ def forward(ctx, caller: RemoteNeuron, dummy: torch.Tensor,
                     return zeros, torch.tensor(bittensor_pb2.ReturnCode.Timeout)
 
                 elif grpc_code == grpc.StatusCode.UNAVAILABLE:
-                    logger.warning('Endpoint unavailable {}', caller.endpoint)
+                    #logger.warning('Endpoint unavailable {}', caller.endpoint)
```
Comment (shibshib) [inline bittensor/dendrite.py:455]: Let's take these loggers out if they're commented.

File: bittensor/dendrites/pkm.py [hunk 4]
```diff
@@ -213,8 +215,8 @@ def _route(self, inputs, query, modality: bittensor_pb2.Modality) -> Tuple[torch
 
         # weights: (torch.LongTensor): weights scattered onto uids per example.
         # weights.shape = [batch_size, metagraph.n]
-        weights = torch.zeros(inputs.shape[0], self.session.metagraph.n).to(self.device)
-        weights.scatter_(1, indices.repeat(batch_size, 1), scores)
+        weights = torch.zeros(inputs.shape[0], self.session.metagraph.n)
```
Comment (shibshib) [inline bittensor/dendrites/pkm.py:218]: Not sure why you took out the `.to(self.device)`, this would break CUDA compatibility on the subsequent `weights.scatter_` call in the next line. Since `indices` is in the GPU and `weights` isn't.

File: bittensor/metagraph.py [hunk 5]
```diff
@@ -214,14 +224,79 @@ def uids(self) -> torch.LongTensor:
...
+        if self.W.shape[0] == 0:
+            return torch.tensor([])
+        else:
+            # S.shape = [self.state.n]
+            # W.shape = [self.state.n, self.state.n]
+            # R.shape = [self.state.n]
+            S = self.S.view(self.state.n, 1)
+            W = torch.transpose(self.W.view(self.state.n, self.state.n), 0, 1)
+            R = torch.matmul(W, S).view(self.state.n)
+        return R
```
Comment (shibshib) [inline bittensor/metagraph.py:289]: The usage of S, W, and R this way is a bit confusing, can we rename the variables to something more descriptive?

File: bittensor/metagraph.py [hunk 6]
```diff
@@ -398,7 +487,7 @@ async def async_emit(self, weights: torch.FloatTensor) -> bool:
         Return:
             included: (bool) true is the weights were set on chain.
         """
-        logger.info('Emit -> {}', weights)
+        #logger.info('Emit -> {}', weights)
```
Comment (shibshib) [inline bittensor/metagraph.py:490]: same as above, let's take out this line if it's unused.

File: bittensor/utils/logging.py [hunk 7]
```diff
@@ -14,9 +14,73 @@
 pd.set_option('display.precision', 2)
 pd.set_option('display.float_format', lambda x: '%.3f' % x)
 
+<<<<<<< HEAD
```
Comment (shibshib) [inline bittensor/utils/logging.py:17]: There's a `<<<<<<<<<<< HEAD` here, did a merge conflict happen somewhere?

File: bittensor/utils/logging.py [hunk 8]
```diff
@@ -14,9 +14,73 @@
...
+        print (df)
+    print('\n')
+
+def log_return_codes(session: Session, history: List[bittensor.synapse.SynapseOutput]):
+    print('Return Codes: \n ')
+=======
 device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
 def calculate_request_sizes_sum(session: Session, history: List[bittensor.synapse.SynapseOutput]):
+>>>>>>> 45917c8b0acefd267a0615bda35b81b5871f7961
```
Comment (shibshib) [inline bittensor/utils/logging.py:83]: merge conflict :)

File: bittensor/axon.py (hunk 1 above)
Comment (unconst) [reply bittensor/axon.py:109]: lol
